cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20AccessTokenController.java

```diff
@@ -81,6 +88,75 @@ private boolean verifyAccessTokenRequest(final HttpServletRequest request) {
             && checkClientSecret(request);
     }
 
+    /**
+     * Check if the credentials exist.
+     *
+     * @param request the HTTP request
+     * @return whether the credentials exist
+     */
+    private boolean checkCredentialsExist(final HttpServletRequest request) {
+        final UsernamePasswordCredentials credential = getCredentials(request);
+        if (StringUtils.isBlank(credential.getUsername())) {
+            logger.error("Missing clientId");
+            return false;
+        } else if (StringUtils.isBlank(credential.getPassword())) {
+            logger.error("Missing secret");
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Check the client secret.
+     *
+     * @param request the HTTP request
+     * @return whether the secret is valid
+     */
+    private boolean checkClientSecret(final HttpServletRequest request) {
+        final UsernamePasswordCredentials credential = getCredentials(request);
+        final String clientId = credential.getUsername();
+        final String clientSecret = credential.getPassword();
+        final OAuthRegisteredService service = OAuthUtils.getRegisteredOAuthService(this.servicesManager, clientId);
+        logger.debug("Found: {} for: {} in secret check", service, clientId);
+
+        if (!StringUtils.equals(service.getClientSecret(), clientSecret)) {
+            logger.error("Wrong client secret for service: {}", service);
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    protected String getClientId(final HttpServletRequest request) {
+        return getCredentials(request).getUsername();
+    }
+
+    /**
+     * Get the client credentials.
+     *
+     * @param request the HTTP request
+     * @return the client credentials
+     */
+    private UsernamePasswordCredentials getCredentials(final HttpServletRequest request) {
+        final String id = request.getParameter(OAuthConstants.CLIENT_ID);
+        final String secret = request.getParameter(OAuthConstants.CLIENT_SECRET);
+        UsernamePasswordCredentials credentials = null;
+        if (StringUtils.isNotBlank(id) && StringUtils.isNotBlank(secret)) {
+            credentials = new UsernamePasswordCredentials(id, secret, null);
+        } else {
+            try {
+                credentials = basicAuthExtractor.extract(new J2EContext(request, null));
+            } catch (final TechnicalException e) {
+                logger.error("Cannot get clientId / secret from header", e);
+            }
+        }
+        if (credentials != null) {
+            return credentials;
+        } else {
+            return new UsernamePasswordCredentials(null, null, null);
```

Comment:
I agree: that's not "pretty", but it saves `null` check on many places. My first design was to return `null`, the ideal would be `Optional` in fact.
I can revert if you want...